library/ssl_tls13_client.c

```diff
@@ -3816,6 +3816,270 @@ static int ssl_hrr_postprocess( mbedtls_ssl_context* ssl,
+static int ssl_new_session_ticket_fetch( mbedtls_ssl_context* ssl,
+                                         unsigned char** dst,
+                                         size_t* dstlen )
+{
+    *dst = ssl->in_msg + mbedtls_ssl_hs_hdr_len( ssl );
+    *dstlen = ssl->in_hslen - mbedtls_ssl_hs_hdr_len( ssl );
+
+    return( 0 );
+}
+#endif /* MBEDTLS_SSL_USE_MPS */
+
+static int ssl_new_session_ticket_parse( mbedtls_ssl_context* ssl,
+                                         unsigned char* buf,
+                                         size_t buflen )
+{
+    int ret;
+    size_t ticket_len, ext_len;
+    unsigned char *ticket;
+    const mbedtls_ssl_ciphersuite_t *suite_info;
+    size_t used = 0, i = 0;
+    int hash_length;
+
+    /*
+     * struct {
+     *    uint32 ticket_lifetime;
+     *    uint32 ticket_age_add;
+     *    opaque ticket_nonce<0..255>;
+     *    opaque ticket<1..2^16-1>;
+     *    Extension extensions<0..2^16-2>;
+     * } NewSessionTicket;
+     *
+     */
+    used += 4   /* ticket_lifetime */
+          + 4   /* ticket_age_add */
+          + 1   /* ticket_nonce length */
+          + 2   /* ticket length */
+          + 2;  /* extension length */
+
+    if( used > buflen )
+    {
+         MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad new session ticket message" ) );
+         return( MBEDTLS_ERR_SSL_BAD_HS_NEW_SESSION_TICKET );
+    }
+
+    /* Ticket lifetime */
+    ssl->session->ticket_lifetime =
+        ( (unsigned) buf[i] << 24 ) | ( (unsigned) buf[i + 1] << 16 ) |
+        ( (unsigned) buf[i + 2] << 8  ) | ( (unsigned) buf[i + 3] << 0 );
+    i += 4;
+
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "ticket->lifetime: %u",
+                                ssl->session->ticket_lifetime ) );
+
+    /* Ticket Age Add */
+    ssl->session->ticket_age_add =
+                     ( (unsigned) buf[i] << 24 ) |
+                     ( (unsigned) buf[i + 1] << 16 ) |
+                     ( (unsigned) buf[i + 2] << 8  ) |
+                     ( (unsigned) buf[i + 3] << 0  );
+    i += 4;
+
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "ticket->ticket_age_add: %u",
+                                ssl->session->ticket_age_add ) );
+
+    /* Ticket Nonce */
+    /* Check if we previously received a ticket already. If we did, then we should
+     * re-use already allocated nonce-space.
+     */
+    if( ssl->session->ticket_nonce != NULL && ssl->session->ticket_nonce_len > 0 )
+    {
+        mbedtls_free( ssl->session->ticket_nonce );
+        ssl->session->ticket_nonce = NULL;
+        ssl->session->ticket_nonce_len = 0;
+    }
+
+    ssl->session->ticket_nonce_len = buf[i];
+    i++;
+
+    used += ssl->session->ticket_nonce_len;
+
+    if( used > buflen )
+    {
+         MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad new session ticket message" ) );
+         return( MBEDTLS_ERR_SSL_BAD_HS_NEW_SESSION_TICKET );
+    }
+
+    if( ssl->session->ticket_nonce_len > 0 )
+    {
+        if( ( ssl->session->ticket_nonce = mbedtls_calloc( 1,
```

Comment:
Could you take a look to see what common nonce lengths are? If they are commonly small and predictable, I would suggest to introduce a compile-time constant for the size of the nonce-buffer, and remove yet another heap allocation.